### PR TITLE
Handle dask arrays

### PR DESCRIPTION
**Pull request by hdsingh**
Now we can create plot for dask arrays.

```python
import xarray as xr
from xrviz.dashboard import Dashboard

ds = xr.tutorial.open_dataset('air_temperature',
                              chunks={'lat': 25, 'lon': 25, 'time': -1})
dash = Dashboard(ds)
dash.show()
```
The above code now runs without error (on clicking `PLOT`). 

In master branch it gives `TypeError: quantile does not work for arrays stored as dask arrays. Load the data via .compute() or .load() prior to calling this method.`

### CODE REVIEW

**Comment by martindurant** [comment]
This is a terrible idea!
The whole idea of a dask-based xarray is that it is too big to fit into memory, and maybe needs to be processed on a cluster. `.compute()` brings the whole thing into memory. So it will be ok for the case that you are using only one slice, but **not** for the whole array. For the latter case at least, I would certainly use min/max instead of percentiles.

(note that the code should be tested anyway - I'm actually not sure whether `.compute()`, the dask method, is the right thing here, instead of xarray's `.values`).

**Comment by hdsingh** [comment]
I will go through the dask docs before attempting to solve this again, to get better understanding.

**Comment by martindurant** [comment]
OK, but ask questions sooner rather than later - dask is a pretty big project, and you are looking specifically at the xarray interface, which hides/augments some of the dask.array functionality.

File: xrviz/dashboard.py [hunk 1]
```diff
@@ -576,6 +575,13 @@ def both_coords_2d_with_same_dims(self):
         return len(x_dims) == len(y_dims) == 2 and sorted(x_dims) == sorted(y_dims)
 
 
+def find_cmap_limits(sel_data):
+    if isinstance(sel_data.data, dask.array.core.Array):
+        return dask.array.percentile(sel_data.data.ravel(), (10, 90)).compute()
+    else:  # if sel_data.data is numpy.ndarray
+        return [float(q) for q in sel_data.quantile([0.1, 0.9])]
+
```
Comment (hdsingh) [inline xrviz/dashboard.py]: @martindurant What do you think about this approach?

**Comment by hdsingh** [comment]
Use of `method=tdigest` in `dask.array.percentile` would require `crick` and `cython` as dependency. Shall it be used?

File: xrviz/dashboard.py [hunk 2]
```diff
@@ -576,6 +576,13 @@ def both_coords_2d_with_same_dims(self):
         return len(x_dims) == len(y_dims) == 2 and sorted(x_dims) == sorted(y_dims)
 
 
+def find_cmap_limits(sel_data):
+    if isinstance(sel_data.data, dask.array.core.Array):
+        return dask.array.percentile(sel_data.data.ravel(), (10, 90)).compute()
+    else:  # if sel_data.data is numpy.ndarray
```
Comment (martindurant) [inline xrviz/dashboard.py]: Note that in the future, this could also be other things like cupy arrays

**Comment by martindurant** [comment]
OK, we can do this for now. Maybe it will change later.

Note that we can test for the existence of crick.TDIgest and use it, if possible. We should not require it. Also, crick does *not* depend on cython, only numpy ( https://github.com/conda-forge/crick-feedstock/blob/master/recipe/meta.yaml#L26 ). Cython is used during building within conda-forge (or pip, if you install from source). If you do not know, I can give a brief introduction into what cython does at our next meeting.

Please test and use crick if possible, and then we can merge this.

I would like you at some point to justify the rounding to 5 places.

File: xrviz/dashboard.py [hunk 3]
```diff
@@ -339,14 +340,13 @@ def create_indexed_graph(self, *args):
...
-            else ([q for q in sel_data.quantile([0.1, 0.9])])
-        )
+            else find_cmap_limits(sel_data))
 
         color_range = {sel_data.name: (c_lim_lower, c_lim_upper)}
 
         if not cmin:  # if user left blank or initial values are empty
-            self.control.style.lower_limit.value = str(c_lim_lower.values.round(5))
-            self.control.style.upper_limit.value = str(c_lim_upper.values.round(5))
+            self.control.style.lower_limit.value = str(round(c_lim_lower, 5))
```
Comment (hdsingh) [inline xrviz/dashboard.py]: Without rounding, the float values can be upto 16 decimal places (`5.454999923706055`) which might not look good, so just for presentation purposes I am rounding off to 5.
This is not necessary and can be removed.

Comment (martindurant) [reply xrviz/dashboard.py]: It may be right, but we should own that it is an arbitrary decision.

Comment (hdsingh) [reply xrviz/dashboard.py]: I will mention about this in docs.

**Comment by hdsingh** [comment]
I have made relevant changes. Please have a look.

File: xrviz/tests/test_dashboard.py [hunk 4]
```diff
@@ -211,3 +211,11 @@ def test_create_taps_and_series_graph_for_2d_coords(dashboard):
...
     dashboard.create_taps_graph(x=-79.232, y=43.273)
     assert isinstance(dashboard.series_graph[0], pn.pane.holoviews.HoloViews)
+
+
+@pytest.mark.skipif(not has_crick_tdigest, reason='crick.tdigest not present')
+def test_find_cmap_limits_with_crick_tdigest():
+    ds = xr.tutorial.open_dataset('air_temperature',
+                                  chunks={'lat': 25, 'lon': 25, 'time': 10})
+    a, b = find_cmap_limits(ds.air)
+    assert (a, b) == (255.38780056044027, 298.5900340551101)
```
Comment (martindurant) [inline xrviz/tests/test_dashboard.py]: These numbers should not the exact like this, precision may differ by system. Probably it is good enough if `abs(expected - found) < 0.1`

Comment (hdsingh) [reply xrviz/tests/test_dashboard.py]: Corrected. Thanks!

**Comment by martindurant** [comment]
OK, going in when it turns green.

**Pull request merged**